File: go/0020E-有效的括号.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-23 15:58:56
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func isValid(s string) bool {
	stack := make([]byte, 0)
	for _, v := range s {
		if v == '(' {
			stack = append(stack, ')')
		} else if v == '[' {
			stack = append(stack, ']')
		} else if v == '{' {
			stack = append(stack, '}')
		} else if len(stack) == 0 {
			return false
		} else if byte(v) == stack[len(stack)-1] {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}
	return len(stack) == 0
}

File: cpp/0020E-有效的括号.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-18 08:53:12
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> t;
        map<char,char> mp;
        mp[']']='[';
        mp['}']='{';
        mp[')']='(';
        for (auto c:s){
            if (c=='(' || c=='{' || c=='[') t.push(c);
            else if (t.empty()) return false;
            else if (t.top()==mp[c]) t.pop();
            else return false;
        }
        return t.empty();
    }
};